runtime/doc/lsp.txt

```diff
@@ -0,0 +1,647 @@
+          \   'build_on_save': v:true,
+          \   'wait_to_build': 0
+          \ }})
+<
+>
+  -- From Lua
+  vim.lsp.add_filetype_config {
+    name = "clangd";
+    filetype = {"c", "cpp"};
+    cmd = "clangd -background-index";
+    capabilities = {
+      offsetEncoding = {"utf-8", "utf-16"};
+    };
+    on_init = vim.schedule_wrap(function(client, result)
+      if result.offsetEncoding then
+        client.offset_encoding = result.offsetEncoding
+      end
+    end)
+  }
+<
+                                                *vim.lsp.copy_filetype_config()*
+vim.lsp.copy_filetype_config({existing_name}, [{override_config}])
+
+  You can use this to copy an existing filetype configuration and change it by
+  specifying {override_config} which will override any properties in the
+  existing configuration. If you don't specify a new unique name with
+  {override_config.name} then it will try to create one and return it.
+
+  Returns:~
+    `name` the new configuration name.
+
+                                         *vim.lsp.get_filetype_client_by_name()*
+vim.lsp.get_filetype_client_by_name({name})
+
+  Use this to look up a client by its name created from
+  |vim.lsp.add_filetype_config()|.
+
+  Returns nil if the client is not active or the name is not valid.
+
+================================================================================
+                                                           *lsp-core-api*
+These are the core api functions for working with clients. You will mainly be
+using |vim.lsp.start_client()| and |vim.lsp.attach_to_buffer()| for operations
+and |vim.lsp.get_client_by_id()| to retrieve a client by its id after it has
+initialized (or {config.on_init}. see below)
+
+                                                        *vim.lsp.start_client()*
+
+vim.lsp.start_client({config})
+
+  The main function used for starting clients.
+  Start a client and initialize it.
+
+  Its arguments are passed via a configuration object {config}.
+ 
+  Mandatory parameters:~
+ 
+  `root_dir`
+    {string} specifying the directory where the LSP server will base
+    as its rootUri on initialization.
+ 
+  `cmd`
+    {string} or {list} which is the base command to execute for the LSP. A
+    string will be run using |'shell'| and a list will be interpreted as a
+    bare command with arguments passed. This is the same as |jobstart()|.
+ 
+  Optional parameters:~
+ 
+  `cmd_cwd`
+    {string} specifying the directory to launch the `cmd` process. This is not
+    related to `root_dir`.
+    By default, |getcwd()| is used.
+ 
+  `cmd_env`
+    {table} specifying the environment flags to pass to the LSP on spawn.
+    This can be specified using keys like a map or as a list with `k=v` pairs
+    or both. Non-string values are coerced to a string.
+    For example:
+      `{ "PRODUCTION=true"; "TEST=123"; PORT = 8080; HOST = "0.0.0.0"; }`
+ 
+  `capabilities`
+    A {table} which will be merged using |vim.deep_merge| with neovim's
```

Comment:
Isn't `deep_merge` an implementation detail that should probably be skipped? Just `merged` seems fine here.

Reply:
I want to make it clear how it will be merged since it's a deeply nested object.